README.md
# Segment-Segment Intersection

![Build Status](https://github.com/pkinney/segseg_ex/actions/workflows/ci.yaml/badge.svg)
[![Hex.pm](https://img.shields.io/hexpm/v/seg_seg.svg)](https://hex.pm/packages/seg_seg)
[![Documentation](https://img.shields.io/badge/documentation-gray)](https://hexdocs.pm/seg_seg)


Calculates intersection type and location for two line segments.

![Classification of segment-segment intersection](http://i.imgbox.com/hO3zHfNR.png)

## Installation

```elixir
defp deps do
  [{:seg_seg, "~> 1.0"}]
end
```

## Usage

**[Full Documentation](https://hexdocs.pm/seg_seg/SegSeg.html)**

The `SegSeg` module provides a function `intersection` that calculates the
intersection between two line segments and returns a tuple with three elements:

1. Boolean `true` if the two segments intersect at all, `false` if they are
   disjoint
2. An atom representing the classification of the intersection:

- `:interior` - the segments intersect at a point that is interior to both
- `:vertex` - the segments intersect at an endpoint of one or both segments
- `:edge` - the segments are parallel, collinear, and overlap for some non-zero
  length
- `:disjoint` - no intersection exists between the two segments

3. A tuple `{x, y}` representing the point of intersection if the intersection
   is classified as `:interior` or `:vertex`, otherwise `nil`.

## Examples

```elixir
SegSeg.intersection({2, -3}, {4, -1}, {2, -1}, {4, -3}) #=> {true, :interior, {3.0, -2.0}}
SegSeg.intersection({-1, 3}, {2, 4}, {-1, 4}, {-1, 5}) #=> {false, :disjoint, nil}
SegSeg.intersection({1, 2}, {3, 0}, {2, 1}, {4, 2}) #=> {true, :vertex, {2, 1}}
SegSeg.intersection({-1, 0}, {0, 2}, {1, 4}, {-1, 0}) #=> {true, :edge, nil}
```

## Float Precision Issues

It is possible that floating point math imprecision can cause incorrect results for certain inputs.  In situations where this may cause issues, an `epsilon` options is available.  When set to `true` intersection comparisons are made with a very small `epsilon` based on the minimum of the lengths of the provided segment times a very small number (currently 0.0000000001). `epsilon` can also be set to a specific number that will be used as the epsilon value. This eliminates most rounding error, but of course could cause false results in certain situations. This currently only effects `:vertex` results but might be expanded to `:edge` in the future.

```elixir
SegSeg.intersection({4, 3}, {4, 7}, {6.05, 9.05}, {3.95, 6.95}) #=> {true, :interior, {4.0, 6.999999999999998}}
SegSeg.intersection({4, 3}, {4, 7}, {6.05, 9.05}, {3.95, 6.95}, epsilon: true) #=> {true, :vertex, {4, 7}}
```

## Tests

```bash
> mix test
```


CHANGELOG.md
# Change Log

## [1.0.0] - 2023-05-30

### Added

- Add `epsilon` option to intersection that, when set to true or a number, applies
  a tolerance to comparison for vertex checks to avoid issues from floating point
  rounding errors.

### Changed

- Update CI to use Github Actions instead of travis-ci
- Update dependencies


test/vertex_test.exs
defmodule VertexTest do
  use ExUnit.Case

  @circle [
    {0, 1},
    {1, 1},
    {1, 0},
    {1, -1},
    {0, -1},
    {-1, -1},
    {-1, 0},
    {-1, 1},
    {1, 2},
    {2, 1},
    {2, -1},
    {1, -2},
    {-1, -2},
    {-2, -1},
    {-2, 1},
    {-1, 2}
  ]

  test "all the vertix intersections around a single point" do
    for b <- @circle, d <- @circle do
      if b == d do
        assert {true, :edge, nil} == SegSegTest.permutations({0, 0}, b, {0, 0}, d)
      else
        assert {true, :vertex, {0, 0}} == SegSegTest.permutations({0, 0}, b, {0, 0}, d)
      end
    end
  end
end


test/seg_seg_test.exs
defmodule SegSegTest do
  use ExUnit.Case
  doctest SegSeg
  import SegSeg

  test "interior intersection" do
    assert {true, :interior, {0.0, 0.0}} = permutations({-1, -1}, {1, 1}, {-1, 1}, {1, -1})
    assert {true, :interior, {0.75, 0.0}} = permutations({-1, 0}, {1, 0}, {0.5, -1}, {1, 1})
    assert {true, :interior, {_, _}} = intersection({-1, -1}, {0, 0.0000001}, {-1, 1}, {1, -1})
  end

  test "vertex" do
    assert {true, :vertex, {0, 0}} = intersection({-1, -1}, {0, 0}, {-1, 1}, {1, -1})
    assert {true, :vertex, {0, 0}} = intersection({0, 0}, {1, 1}, {0, 0}, {-1, -1})
    assert {true, :vertex, {0.5, 0}} = permutations({-1, 0}, {1, 0}, {0.5, 0}, {1, 1})
    assert {true, :vertex, {2, 1}} = permutations({1, 2}, {3, 0}, {2, 1}, {4, 2})
    assert {true, :vertex, {80, 50}} = permutations({20, 20}, {80, 50}, {80, 50}, {110, 65})
    assert {true, :vertex, {20, 20}} = permutations({20, 20}, {80, 140}, {20, 20}, {20, 180})
  end

  test "vertex (epsilon mode)" do
    # This is a known-bad set of inputs that produce rounding error and return incorrect results
    # only when an epsilon is given.
    assert {true, :interior, _} =
             intersection({4, 3}, {4, 7}, {6.05, 9.05}, {3.95, 6.95}, epsilon: false)

    assert {true, :vertex, {4, 7}} =
             intersection({4, 3}, {4, 7}, {6.05, 9.05}, {3.95, 6.95}, epsilon: true)
  end

  test "edge" do
    assert {true, :edge, nil} = intersection({20, 20}, {80, 50}, {80, 50}, {50, 35})
    assert {true, :edge, nil} = permutations({0, 0}, {1, 1}, {0, 0}, {2, 2})
    assert {true, :edge, nil} = intersection({-1, 0}, {0, 2}, {1, 4}, {-1, 0})
    assert {true, :edge, nil} = permutations({-1, 0}, {1, 0}, {0.5, 0}, {1.5, 0})
    assert {true, :edge, nil} = permutations({-1, 0}, {1, -1.5}, {-1, 0}, {1, -1.5})
  end

  test "disjoint, not parallel" do
    assert {false, :disjoint, nil} = permutations({-1, 0}, {1, 0}, {0.5, -1}, {1, -1.5})
    assert {false, :disjoint, nil} = permutations({-1, 0}, {0, 0}, {1, 1}, {1, -1})
    assert {false, :disjoint, nil} = permutations({1, 3}, {-1, 0}, {-1, -1}, {-2, 0})
  end

  test "disjoint, parallel" do
    assert {false, :disjoint, nil} = permutations({20, 20}, {80, 50}, {25, 45}, {55, 60})
  end

  test "disjoint, collinear" do
    assert {false, :disjoint, nil} = permutations({20, 20}, {80, 50}, {75, 45}, {105, 60})
    assert {false, :disjoint, nil} = permutations({20, 20}, {30, 25}, {40, 30}, {50, 35})
    assert {false, :disjoint, nil} = permutations({20, 20}, {80, 50}, {90, 55}, {110, 65})
  end

  def permutations(a, b, c, d) do
    result = intersection(a, b, c, d)
    assert result == intersection(b, a, c, d)
    assert result == intersection(a, b, d, c)
    assert result == intersection(b, a, d, c)
    assert result == intersection(c, d, a, b)
    assert result == intersection(d, c, a, b)
    assert result == intersection(c, d, b, a)
    assert result == intersection(d, c, b, a)
    assert result == intersection(a, b, c, d, epsilon: true)
    assert result == intersection(b, a, c, d, epsilon: true)
    assert result == intersection(a, b, d, c, epsilon: true)
    assert result == intersection(b, a, d, c, epsilon: true)
    assert result == intersection(c, d, a, b, epsilon: true)
    assert result == intersection(d, c, a, b, epsilon: true)
    assert result == intersection(c, d, b, a, epsilon: true)
    assert result == intersection(d, c, b, a, epsilon: true)
    result
  end
end


test/disjiont_test.exs
defmodule DisjointTest do
  use ExUnit.Case

  @segments [
    {{0, 0}, {1, 1}},
    {{1, 4}, {2, 4}},
    {{1, 6}, {3, 6}},
    {{2, 2}, {2, 3}},
    {{3, 1}, {3, 2}},
    {{3, 3}, {5, 5}},
    {{3, 4}, {3, 5}},
    {{4, 1}, {6, 3}},
    {{4, 2}, {4, 3}},
    {{4, 5}, {4, 6}},
    {{5, 3}, {5, 4}},
    {{5, 6}, {5, 7}},
    {{6, 5}, {7, 5}},
    {{6, 6}, {8, 8}},
    {{7, 1}, {7, 2}},
    {{7, 3}, {7, 4}},
    {{7, 6}, {9, 4}},
    {{9, 5}, {11, 3}},
    {{10, 3}, {11, 2}},
    {{10, 5}, {11, 5}}
  ]

  test "all the disjoint" do
    for {ax, ay} <- @segments, {bx, by} <- @segments do
      if ax == bx && ay == by do
        assert {true, :edge, nil} == SegSegTest.permutations(ax, ay, bx, by)
      else
        assert {false, :disjoint, nil} == SegSegTest.permutations(ax, ay, bx, by)
      end
    end
  end
end


mix.exs
defmodule SegSeg.Mixfile do
  use Mix.Project

  def project do
    [
      app: :seg_seg,
      version: "1.0.0",
      elixir: "~> 1.2",
      description: description(),
      package: package(),
      build_embedded: Mix.env() == :prod,
      start_permanent: Mix.env() == :prod,
      test_coverage: [tool: ExCoveralls],
      preferred_cli_env: [coveralls: :test],
      dialyzer: [plt_add_apps: [:mix]],
      deps: deps(),
      aliases: aliases()
    ]
  end

  def application do
    [applications: [:logger, :vector]]
  end

  defp deps do
    [
      {:vector, "~> 1.0"},
      {:earmark, "~> 1.0", only: :dev},
      {:ex_doc, "~> 0.16", only: :dev},
      {:excoveralls, "~> 0.4", only: :test},
      {:credo, "~> 1.6", only: [:dev, :test], runtime: false},
      {:dialyxir, "~> 1.2", only: [:dev], runtime: false}
    ]
  end

  defp description do
    """
    Segment-Segment intersection point and classification
    """
  end

  defp package do
    [
      files: ["lib", "mix.exs", "README*"],
      maintainers: ["Powell Kinney"],
      licenses: ["MIT"],
      links: %{"GitHub" => "https://github.com/pkinney/segseg_ex"}
    ]
  end

  defp aliases do
    [
      validate: [
        "clean",
        "compile --warnings-as-error",
        "format --check-formatted",
        "credo",
        "dialyzer"
      ]
    ]
  end
end


.credo.exs
# This file contains the configuration for Credo and you are probably reading
# this after creating it with `mix credo.gen.config`.
#
# If you find anything wrong or unclear in this file, please report an
# issue on GitHub: https://github.com/rrrene/credo/issues
#
%{
  #
  # You can have as many configs as you like in the `configs:` field.
  configs: [
    %{
      #
      # Run any config using `mix credo -C <name>`. If no config name is given
      # "default" is used.
      #
      name: "default",
      #
      # These are the files included in the analysis:
      files: %{
        #
        # You can give explicit globs or simply directories.
        # In the latter case `**/*.{ex,exs}` will be used.
        #
        included: [
          "lib/",
          "src/",
          "test/",
          "web/",
          "apps/*/lib/",
          "apps/*/src/",
          "apps/*/test/",
          "apps/*/web/"
        ],
        excluded: [~r"/_build/", ~r"/deps/", ~r"/node_modules/"]
      },
      #
      # Load and configure plugins here:
      #
      plugins: [],
      #
      # If you create your own checks, you must specify the source files for
      # them here, so they can be loaded by Credo before running the analysis.
      #
      requires: [],
      #
      # If you want to enforce a style guide and need a more traditional linting
      # experience, you can change `strict` to `true` below:
      #
      strict: false,
      #
      # To modify the timeout for parsing files, change this value:
      #
      parse_timeout: 5000,
      #
      # If you want to use uncolored output by default, you can change `color`
      # to `false` below:
      #
      color: true,
      #
      # You can customize the parameters of any check by adding a second element
      # to the tuple.
      #
      # To disable a check put `false` as second element:
      #
      #     {Credo.Check.Design.DuplicatedCode, false}
      #
      checks: %{
        enabled: [
          #
          ## Consistency Checks
          #
          {Credo.Check.Consistency.ExceptionNames, []},
          {Credo.Check.Consistency.LineEndings, []},
          {Credo.Check.Consistency.ParameterPatternMatching, []},
          {Credo.Check.Consistency.SpaceAroundOperators, []},
          {Credo.Check.Consistency.SpaceInParentheses, []},
          {Credo.Check.Consistency.TabsOrSpaces, []},

          #
          ## Design Checks
          #
          # You can customize the priority of any check
          # Priority values are: `low, normal, high, higher`
          #
          {Credo.Check.Design.AliasUsage,
           [priority: :low, if_nested_deeper_than: 2, if_called_more_often_than: 0]},
          # You can also customize the exit_status of each check.
          # If you don't want TODO comments to cause `mix credo` to fail, just
          # set this value to 0 (zero).
          #
          {Credo.Check.Design.TagTODO, [exit_status: 2]},
          {Credo.Check.Design.TagFIXME, []},

          #
          ## Readability Checks
          #
          {Credo.Check.Readability.AliasOrder, []},
          {Credo.Check.Readability.FunctionNames, []},
          {Credo.Check.Readability.LargeNumbers, []},
          {Credo.Check.Readability.MaxLineLength, [priority: :low, max_length: 120]},
          {Credo.Check.Readability.ModuleAttributeNames, []},
          {Credo.Check.Readability.ModuleDoc, []},
          {Credo.Check.Readability.ModuleNames, []},
          {Credo.Check.Readability.ParenthesesInCondition, []},
          {Credo.Check.Readability.ParenthesesOnZeroArityDefs, []},
          {Credo.Check.Readability.PipeIntoAnonymousFunctions, []},
          {Credo.Check.Readability.PredicateFunctionNames, []},
          {Credo.Check.Readability.PreferImplicitTry, []},
          {Credo.Check.Readability.RedundantBlankLines, []},
          {Credo.Check.Readability.Semicolons, []},
          {Credo.Check.Readability.SpaceAfterCommas, []},
          {Credo.Check.Readability.StringSigils, []},
          {Credo.Check.Readability.TrailingBlankLine, []},
          {Credo.Check.Readability.TrailingWhiteSpace, []},
          {Credo.Check.Readability.UnnecessaryAliasExpansion, []},
          {Credo.Check.Readability.VariableNames, []},
          {Credo.Check.Readability.WithSingleClause, []},

          #
          ## Refactoring Opportunities
          #
          {Credo.Check.Refactor.Apply, []},
          {Credo.Check.Refactor.CondStatements, []},
          # @pkinney
          # {Credo.Check.Refactor.CyclomaticComplexity, []},
          {Credo.Check.Refactor.FunctionArity, []},
          {Credo.Check.Refactor.LongQuoteBlocks, []},
          {Credo.Check.Refactor.MatchInCondition, []},
          {Credo.Check.Refactor.MapJoin, []},
          {Credo.Check.Refactor.NegatedConditionsInUnless, []},
          {Credo.Check.Refactor.NegatedConditionsWithElse, []},
          {Credo.Check.Refactor.Nesting, []},
          {Credo.Check.Refactor.UnlessWithElse, []},
          {Credo.Check.Refactor.WithClauses, []},
          {Credo.Check.Refactor.FilterFilter, []},
          {Credo.Check.Refactor.RejectReject, []},
          {Credo.Check.Refactor.RedundantWithClauseResult, []},

          #
          ## Warnings
          #
          {Credo.Check.Warning.ApplicationConfigInModuleAttribute, []},
          {Credo.Check.Warning.BoolOperationOnSameValues, []},
          {Credo.Check.Warning.ExpensiveEmptyEnumCheck, []},
          {Credo.Check.Warning.IExPry, []},
          {Credo.Check.Warning.IoInspect, []},
          {Credo.Check.Warning.OperationOnSameValues, []},
          {Credo.Check.Warning.OperationWithConstantResult, []},
          {Credo.Check.Warning.RaiseInsideRescue, []},
          {Credo.Check.Warning.SpecWithStruct, []},
          {Credo.Check.Warning.WrongTestFileExtension, []},
          {Credo.Check.Warning.UnusedEnumOperation, []},
          {Credo.Check.Warning.UnusedFileOperation, []},
          {Credo.Check.Warning.UnusedKeywordOperation, []},
          {Credo.Check.Warning.UnusedListOperation, []},
          {Credo.Check.Warning.UnusedPathOperation, []},
          {Credo.Check.Warning.UnusedRegexOperation, []},
          {Credo.Check.Warning.UnusedStringOperation, []},
          {Credo.Check.Warning.UnusedTupleOperation, []},
          {Credo.Check.Warning.UnsafeExec, []}
        ],
        disabled: [
          #
          # Checks scheduled for next check update (opt-in for now, just replace `false` with `[]`)

          #
          # Controversial and experimental checks (opt-in, just move the check to `:enabled`
          #   and be sure to use `mix credo --strict` to see low priority checks)
          #
          {Credo.Check.Consistency.MultiAliasImportRequireUse, []},
          {Credo.Check.Consistency.UnusedVariableNames, []},
          {Credo.Check.Design.DuplicatedCode, []},
          {Credo.Check.Design.SkipTestWithoutComment, []},
          {Credo.Check.Readability.AliasAs, []},
          {Credo.Check.Readability.BlockPipe, []},
          {Credo.Check.Readability.ImplTrue, []},
          {Credo.Check.Readability.MultiAlias, []},
          {Credo.Check.Readability.NestedFunctionCalls, []},
          {Credo.Check.Readability.SeparateAliasRequire, []},
          {Credo.Check.Readability.SingleFunctionToBlockPipe, []},
          {Credo.Check.Readability.SinglePipe, []},
          {Credo.Check.Readability.Specs, []},
          {Credo.Check.Readability.StrictModuleLayout, []},
          {Credo.Check.Readability.WithCustomTaggedTuple, []},
          # @pkinney
          # {Credo.Check.Refactor.ABCSize, []},
          {Credo.Check.Refactor.AppendSingleItem, []},
          {Credo.Check.Refactor.DoubleBooleanNegation, []},
          {Credo.Check.Refactor.FilterReject, []},
          {Credo.Check.Refactor.IoPuts, []},
          {Credo.Check.Refactor.MapMap, []},
          {Credo.Check.Refactor.ModuleDependencies, []},
          {Credo.Check.Refactor.NegatedIsNil, []},
          {Credo.Check.Refactor.PipeChainStart, []},
          {Credo.Check.Refactor.RejectFilter, []},
          {Credo.Check.Refactor.VariableRebinding, []},
          {Credo.Check.Warning.LazyLogging, []},
          {Credo.Check.Warning.LeakyEnvironment, []},
          {Credo.Check.Warning.MapGetUnsafePass, []},
          {Credo.Check.Warning.MixEnv, []},
          {Credo.Check.Warning.UnsafeToAtom, []}

          # {Credo.Check.Refactor.MapInto, []},

          #
          # Custom checks can be created using `mix credo.gen.check`.
          #
        ]
      }
    }
  ]
}


config/config.exs
# This file is responsible for configuring your application
# and its dependencies with the aid of the Mix.Config module.
import Config

# This configuration is loaded before any dependency and is restricted
# to this project. If another project depends on this project, this
# file won't be loaded nor affect the parent project. For this reason,
# if you want to provide default values for your application for
# 3rd-party users, it should be done in your "mix.exs" file.

# You can configure for your application as:
#
#     config :seg_seg, key: :value
#
# And access this configuration in your application as:
#
#     Application.get_env(:seg_seg, :key)
#
# Or configure a 3rd-party app:
#
#     config :logger, level: :info
#

# It is also possible to import configuration files, relative to this
# directory. For example, you can emulate configuration per environment
# by uncommenting the line below and defining dev.exs, test.exs and such.
# Configuration from the imported file will override the ones defined
# here (which is why it is important to import them last).
#
#     import_config "#{Mix.env}.exs"


lib/seg_seg.ex
defmodule SegSeg do
  @moduledoc ~S"""
  Calculates the type of relationship between two line segments **AB** and
  **CD** and the location of intersection (if applicable).

  ![Classification of segment-segment intersection](http://i.imgbox.com/hO3zHfNR.png)

  """

  @type point :: {number, number}
  @type intersection_type ::
          :interior
          | :disjoint
          | :edge
          | :vertex
  @type intersection_result :: {boolean, intersection_type, point | nil}

  # When the `epsilon` option is passed as `true` or a number, certain comparisons are broadened
  # based on an epsilon value. 
  # If `true` is passed, the eps value is defined as the smaller of the distances
  # of each segment multiplied by the `@eps_factor` below.
  @eps_factor 0.000000001

  @doc ~S"""
  Returns a tuple representing the segment-segment intersection with three
  elements:

  1. Boolean `true` if the two segments intersect at all, `false` if they are
     disjoint
  2. An atom representing the classification of the intersection:
    * `:interior` - the segments intersect at a point that is interior to both
    * `:vertex` - the segments intersect at an endpoint of one or both segments
    * `:edge` - the segments are parallel, collinear, and overlap for some non-zero length
    * `:disjoint` - no intersection exists between the two segments
  3. A tuple `{x, y}` representing the point of intersection if the intersection
     is classified as `:interior` or `:vertex`, otherwise `nil`.

  ## Float Precision Issues

  It is possible that floating point math imprecision can cause incorrect results for
  certain inputs.  In situations where this may cause issues, an `epsilon` options is
  available.  When set to `true` intersection comparisons are made with a very small `epsilon` based on the minimum
  of the lengths of the provided segment times a very small number (currently 0.0000000001). `epsilon` can also be set to a specific number that will be used as the epsilon value.
  This eliminates most rounding error, but of course could cause false results in certain
  situations. This currently only effects `:vertex` results but might be expanded to `:edge`
  in the future.

  ```elixir
  SegSeg.intersection({4, 3}, {4, 7}, {6.05, 9.05}, {3.95, 6.95}) #=> {true, :interior, {4.0, 6.999999999999998}}
  SegSeg.intersection({4, 3}, {4, 7}, {6.05, 9.05}, {3.95, 6.95}, epsilon: true) #=> {true, :vertex, {4, 7}}
  ```

  ## Examples

      iex> SegSeg.intersection({2, -3}, {4, -1}, {2, -1}, {4, -3})
      {true, :interior, {3.0, -2.0}}
      iex> SegSeg.intersection({-1, 3}, {2, 4}, {-1, 4}, {-1, 5})
      {false, :disjoint, nil}
      iex> SegSeg.intersection({-1, 0}, {0, 2}, {0, 2}, {1, -1})
      {true, :vertex, {0, 2}}
      iex> SegSeg.intersection({-1, 0}, {0, 2}, {1, 4}, {-1, 0})
      {true, :edge, nil}

      # This doesn't behave as expected because of floating point error.
      iex> SegSeg.intersection({4, 3}, {4, 7}, {6.05, 9.05}, {3.95, 6.95})
      {true, :interior, {4.0, 6.999999999999998}}

      # Adding the default epsilon causes this to act more as expected.
      iex> SegSeg.intersection({4, 3}, {4, 7}, {6.05, 9.05}, {3.95, 6.95}, epsilon: true)
      {true, :vertex, {4, 7}}

      # Specifying a very small epsilon shows the floating point error again.
      iex> SegSeg.intersection({4, 3}, {4, 7}, {6.05, 9.05}, {3.95, 6.95}, epsilon: 0.00000000000000000001)
      {true, :interior, {4.0, 6.999999999999998}}
  """
  @spec intersection(point, point, point, point, keyword()) :: intersection_result
  def intersection(a, b, c, d, options \\ []) do
    if envelope_check(a, b, c, d) do
      denom = calc_denom(a, b, c, d)
      eps = calc_eps(a, b, c, d, options)
      do_intersection(a, b, c, d, denom, eps)
    else
      {false, :disjoint, nil}
    end
  end

  defp do_intersection(a, b, c, d, denom, _) when denom == 0, do: parallel_int(a, b, c, d)
  defp do_intersection(a, _, c, d, _, _) when a == c or a == d, do: {true, :vertex, a}
  defp do_intersection(_, b, c, d, _, _) when b == c or b == d, do: {true, :vertex, b}

  # credo:disable-for-next-line Credo.Check.Refactor.CyclomaticComplexity
  defp do_intersection({ax, ay}, {bx, by}, {cx, cy}, {dx, dy}, denom, eps) do
    s = (ax * (dy - cy) + cx * (ay - dy) + dx * (cy - ay)) / denom
    t = -(ax * (cy - by) + bx * (ay - cy) + cx * (by - ay)) / denom

    t_zero = t >= -eps && t <= eps
    t_one = t >= 1.0 - eps && t <= 1.0 + eps
    s_zero = s >= -eps && s <= eps
    s_one = s >= 1.0 - eps && s <= 1.0 + eps

    cond do
      (t_zero || t_one) && collinear_not_between({ax, ay}, {bx, by}, {cx, cy}) ->
        {false, :disjoint, nil}

      (t_zero || t_one) && collinear_not_between({ax, ay}, {bx, by}, {dx, dy}) ->
        {false, :disjoint, nil}

      (s_zero || s_one) && collinear_not_between({cx, cy}, {dx, dy}, {ax, ay}) ->
        {false, :disjoint, nil}

      (s_zero || s_one) && collinear_not_between({cx, cy}, {dx, dy}, {bx, by}) ->
        {false, :disjoint, nil}

      s_zero ->
        {true, :vertex, {ax, ay}}

      s_one ->
        {true, :vertex, {bx, by}}

      t_zero ->
        {true, :vertex, {cx, cy}}

      t_one ->
        {true, :vertex, {dx, dy}}

      s > 0.0 && s < 1.0 && t > 0.0 && t < 1.0 ->
        {true, :interior, {ax + s * (bx - ax), ay + s * (by - ay)}}

      true ->
        {false, :disjoint, nil}
    end
  end

  defp calc_eps(a, b, c, d, options) do
    case Keyword.get(options, :epsilon, false) do
      false -> 0.0
      true -> eps_factor(a, b, c, d)
      num when is_number(num) -> num
    end
  end

  defp eps_factor({ax, ay}, {bx, by}, {cx, cy}, {dx, dy}) do
    ab_dist = (ax - bx) * (ax - bx) + (ay - by) * (ay - by)
    cd_dist = (cx - dx) * (cx - dx) + (cy - dy) * (cy - dy)
    :math.sqrt(min(ab_dist, cd_dist)) * @eps_factor
  end

  defp calc_denom({ax, ay}, {bx, by}, {cx, cy}, {dx, dy}) do
    ax * (dy - cy) + bx * (cy - dy) + dx * (by - ay) + cx * (ay - by)
  end

  # credo:disable-for-next-line Credo.Check.Refactor.CyclomaticComplexity
  defp envelope_check({ax, ay}, {bx, by}, {cx, cy}, {dx, dy}) do
    cond do
      ax < cx && ax < dx && bx < cx && bx < dx -> false
      ax > cx && ax > dx && bx > cx && bx > dx -> false
      ay < cy && ay < dy && by < cy && by < dy -> false
      ay > cy && ay > dy && by > cy && by > dy -> false
      true -> true
    end
  end

  # credo:disable-for-next-line Credo.Check.Refactor.CyclomaticComplexity
  defp parallel_int(a, b, c, d) do
    cond do
      !collinear(a, b, c) -> {false, :disjoint, nil}
      a == c && (between(a, b, d) || between(a, d, b)) -> {true, :edge, nil}
      a == d && (between(a, b, c) || between(a, c, b)) -> {true, :edge, nil}
      b == c && (between(a, b, d) || between(b, d, a)) -> {true, :edge, nil}
      b == d && (between(a, b, c) || between(b, c, a)) -> {true, :edge, nil}
      a == c || a == d -> {true, :vertex, a}
      b == c || b == d -> {true, :vertex, b}
      true -> {true, :edge, nil}
    end
  end

  defp collinear_not_between(a, b, c) do
    collinear(a, b, c) && !between(a, b, c)
  end

  defp collinear({ax, ay}, {bx, by}, {cx, cy}) do
    Vector.cross({ax - cx, ay - cy}, {bx - cx, by - cy}) == {0, 0, 0}
  end

  defp between({ax, ay}, {bx, by}, {_, py}) when ax == bx,
    do: (ay <= py && py <= by) || (ay >= py && py >= by)

  defp between({ax, _}, {bx, _}, {px, _}), do: (ax <= px && px <= bx) || (ax >= px && px >= bx)
end
